### ISSUE TITLE
Integration Testing

### ISSUE THREAD

**Issue by lawrencegripper** [lawrencegripper/azbrowse#81]
Related to #22 

@jjcollinge @martinpeck @stuartleeks - I need your brains!

I REALLY want a set of reasonable integration tests which will run through common operations and check a couple of things. 

Currently I'm thinking this would have to be some kind of test rig which creates a group of resources (TF?) in a known Azure Sub and then launch an instance and *drives* the UI of a started instance of AzBrowse by pushing keys to it over `stdin`. 

### Harder bits

1. It would need to be robust enough to be insensitive to UI changes (granted it can't survive massive ones but as long as up/down/enter work I'd expect it to work). 
2. How easy is it to drive a terminal this way and check the output? Does it need some hooks for the test harness to get the resulting output without screen scraping - how do we do this? Have a special running case which runs GRPC/HTTP server which lets the test harness see the `TreeItems` and `ItemContent`?  

### Operations - Must

- Browse Deployment
- Browse RG
- Browse VM
- Browse Website
- Browse Activity logs

### Operations - Maybe

- Restart VM with actions
- List storage keys
- Delete an item
- Edit an item
- Status indicators

### Checks during an operation

- Does it crash?
- Does it display the right info? (maybe not all but a few key fields)
- Track the time and set max execution time for operation to prevent perf issues being introduced
- Track startup time of app

**Comment by jjcollinge** [comment]
I think it may make sense to start the testing a more granular level for this.
A reasonably low hanging fruit would be to have a test for each handler which is agnostic to the key mappings or GUI.
You could then have tests for the bindings/mappings to check they invoke the correct handler.
Each view could have a test suite which checks it renders and updates correctly when its methods are called.
I think this will probably give us enough test coverage to catch the majority of code related issues.
Then I think we should add behavioural and performance tests for startup/shutdown, soaking and key user journies. Like you say this could be rigged up so that we have a generic way of booting azbrowse in a known state, automating user interactions (keyboard driving) and capturing the required data.

**Comment by lawrencegripper** [comment]
I like this plan, I can see the handlers working well as they're nicely isolated from `gocui` so will give that a go. 

With the `views` I hit some issues and I'm not sure the best way to go. 

So in the #85 I tried to add a test like this 

```golang
func Test_AddPendingDelete(t *testing.T) {

	g, err := gocui.NewGui(gocui.OutputNormal)
	if err != nil {
		t.Error(err)
		t.FailNow()
	}
	go func() {
		err := g.MainLoop()
		t.Error(err)
	}()
	notView := NewNotificationWidget(0, 0, 45, false, g)
	g.SetManager(notView)

	AddPendingDelete("s1", "http://delete/s1")
	AddPendingDelete("s2", "http://delete/s2")
	AddPendingDelete("s3", "http://delete/s3")

	builder := &strings.Builder{}
	err = notView.layoutInternal(builder)
	if err != nil {
		t.Error(err)
	}

	expected := "some payout"

	if builder.String() != expected {
		t.Errorf("Expected: %s Got: %s", expected, builder.String())
	}
}
```
 
`layoutInternal` is the `Layout` method from the view but taken out of `gocui` so it takes in an `io.writer`.

The problem is the pervasive nature of `gocui` in the code so `AddPendingDelete` uses it for `Size()` and `Update()`. As we're in an automated test and not a console these fail as there is no `/dev/tty` device for `gocui -> termbox` to use to write to the console. 

I guess I see a couple of options for testing the views:

1. Migrate away from `gocui` or fork it to allow it to have a `file` target for testing not just `/dev/tty`. I really don't know how painful this would be ... my guess pretty painful
2. Try and wrap `gocui` into an abstraction that we can mock for testing. This feels like the best route I can think of. 

I'd also look to mock out `armclient` too so we can assert that, for example, `confirmDelete` sends the expected delete calls to the API. 

What do you think?

**Comment by lawrencegripper** [comment]
Scratch that I managed to get this working by looking at another `gocui` project which has some ui testing https://github.com/gchaincl/httplab/blob/master/ui/ui_test.go

**Comment by lawrencegripper** [comment]
I've attempted to add a decent set of tests to the new work in #85 testing both the view logic and the `armclient` calls made. 

Would be great to get peoples thoughts on the approach https://github.com/lawrencegripper/azbrowse/pull/85/commits/1676e20d93e02380798178f63cea5065a00ba6b8

**Comment by lawrencegripper** [comment]
For the time being I can't run the `tty` based tests in CircleCI. I've split this out into a `make integration` test which can be run locally.

**Comment by lawrencegripper** [comment]
Managed to get a full XServer instance running under AzureDevOps with a dummy display so I can launch XTerm and run rull test suite ... took an age but I'm happy!

**Comment by jjcollinge** [comment]
Sounds great, will take a look when I get time - are you migrating the pipelines over to ADO from Circle or using both?

**Comment by lawrencegripper** [comment]
As the existing pipeline works well and they both cover different bit's I was intending to keep both, added plus is then the integration runs in parallel so not so long to wait either.

**Comment by lawrencegripper** [comment]
Think we're in a good place with this now.
